### DISCUSSION TITLE
Working together on a helm chart!

### DISCUSSION THREAD

**Discussion by EugenMayer** [buanet/ioBroker.docker#417] [General Discussions / Allgemeine Diskussionen]
### Description / Beschreibung

Hey there,

visiting https://docs.buanet.de/iobroker-docker-image/docs/ due to my interest in running ioBroker in a container env. i stumbeled uppon the guy with the role "Smarthome on docker" :)

I'am currently build out my homelab, switched from plain containerd/podman+docker-compose to k8s entirely. Next step is to run ioBroker in k8s and AFAICS, there is no helm chart yet, so i'am going to make one most probably.

That in mind, i asked myself if you are interested in working together on this adventure - my goals are

- ensure one can run with out-discovery and utilizing VLAN based networking (home automation network isolation) - this is what i plan to use multus for
- ensure to run ioBroker as slim as possible, keep services like redis and other things in seperate pods/containers (scaling)
- scaling via multimode - since i have a 1+ node cluster anyway, this helps spreading the load (which ioBroker can be very good at creating :) )
- i would avoid containerizing ioBroker myself is possibel , this is where this project might be interesting to use with

The last point might need a bit of adoption (more or less introducing ENV vars if any - but there is a lot here already, so might not be needed at all).

What are your thoughts about all of this?

Thanks for this little gem of a project :)

**Comment by EugenMayer** [comment]
Created a first (working) draft here https://github.com/EugenMayer/helm-charts/tree/task/iobroker/charts/iobroker

Taken care of:
- env based config
- volume / persistence
- ingress

TODO: 
- networking (multicast/broadcast) VLAN support via multus as an option

**Comment by buanet** [comment]
If I take a look into the ioBroker community, I only see a very very small target group for running ioBroker inside a kubernetes cluster... So, for me, there is no need to put time into this to make something like a helm-chart available to the community.

As I'm not able to put more time into this project, I will only focus to things with bigger target groups in the community.
That doesn't say, that I won't support your effort. 

I'll convert this issue to an discussion. So if there is somebody out there waiting for a helm-chart her/ she can easily join you. 

Regards,
André

**Reply by EugenMayer** [reply]
I understand the audience is smallish. I would tend to rather think this changes in the future (as it did for home-assistant too).

Still, not in need of actual help here - i will do the work. It will most probably come down to advice what people go for usually with ioBroker, since i'am not a heavy user yet. But this can grow over time.

Should i inform you here if the chart can be judged as "finished for deployments" or are you generally not interested in it. Not trying to force it on you in any way.

**Reply by buanet** [reply]
Ya, this is a good place for updates. As I have no setup to test it yet, I can only take a "theoretical" look at it. 😄
We will keep this discussion open, so if there are people searching or asking for it we can use this as "meeting point". 

If you need some adjustments within the docker image we can discuss this here too.
If you like, we have a "iobroker image" Discord channel for "beta testing" at the ioBroker Discord Server. 

Regards,
André

**Reply by Apollon77** [reply]
Hey, I think  this is a interesting idea, but also challenging.

A collegue of mine also did something like this some time ago ... here is what he did send me (no idea if up to date tbh) https://gitlab.com/psych0d0g/k8s_cluster/-/blob/main/deployments/iobroker/helmrelease.yaml?ref_type=heads

The main challeng in my eyes is that ioBroker itself is currently not "scalable" via default K8s scaling because the adfapter instances are bound to hosts with a specific name and so it will not really work to say" scale up". In fact each "iobroker server" you define can only have one pod and needs to be defined with a hard name and such ... Then using redis all should be fine.

But it is a rather spüecial setup for K8s in my eyes.

**Comment by UncleSamSwiss** [comment]
@EugenMayer I have ioBroker running in k3s at home as well. I'm not using a helm chart but rather Terraform, but that's not a big difference in functionality (just a completely different language).

For now I don't have a cluster (i.e. only a single node running on my NAS). But here it gets interesting: I have 4 Raspis running ioBroker as well to handle local USB and I2C devices - all is connected as a multi-host setup.

This is all running fine and I could have the Raspis joining the cluster to have the benefits of a real Kubernetes cluster. But as @Apollon77 also wrote, there are some basic issues with this:

1. ioBroker is not scalable, an adapter can never run twice at the same time
2. Adapters are run by js-controller, so the smallest unit to map to a pod is not an adapter, but an entire js-controller with (potentially) a lot of adapters
3. Quite a few adapters need an open port on the local network (web, admin, mqtt,...), all these ports should be LoadBalancer services in Kubernetes, otherwise there can't be a failover in case a node goes down
4. Some adapters need to run on a given host because they are accessing hardware (USB, I2C, RS232, ...)

The solution for all of these issues is currently a thought experiment in my head only, but it involves:
- replacing parts of js-controller by the k8s control plane using something similar to an Operator
- a special docker image for each adapter (obviously generated automatically)
- depending on the adapter, a special configuration for additional k8s resources (like LoadBalancer, Service, ...)
- helm chart for the basic installation of the "ioBroker Operator" and ioBroker.admin
- some nice tooling to install all of this in a k3s without regular users needing to setup all themselves

So, to sum up: yes, I know, ioBroker can and does run in Kubernetes, but it currently makes little sense outside the circle of people who know Kubernetes well and in most cases, the setup is very specific to the use case. To have a valuable product "for the masses", we need to overcome the above issues by making ioBroker "Kubernetes native", which is a big endeavor.

**Reply by Apollon77** [reply]
I‘m not sure if it is that easy to replace party of the js-controller, but inteesting thought. In fact the host management need to be adjusted to be rather static. 

In general I had the idea of an higher availability adapter some time ago that would hold the config which instances need local resources and which could be moved around when a node fails. The js-controller already have a master selection internally when you have multiple hosts. so idea was to use this to select „the master controller“ who would orchestrate such a detection and moving around instances by assigning different hosts. 
but time is always too short ;-)

**Reply by EugenMayer** [reply]
@UncleSamSwiss 

Running k3s too here, deploying the actuall not via tf on proxmox. Also using tf to deploy and configure my all charts to (seeding values yamls) and for the usual ingress/pvc parts that i also handle outside charts (including network security policies) - so not new to this, but a helm chart is a must out of a lot of reasons, also with TF.

Yes operators are the usual option to control such an environment, i'am running PGO and mongod-community-cluster myself here, which both are based on operators. But writing an operator usually involves a lot of detail and also time, which is currently out of scope for me.

I finished the helm chart for now, i based it on TrueCharts for maximum flexibilityand this my first goal. Now i first will get into the weeds getting value out of ioBroker, before i continue to go for any HA Setups.

But on my roadmap, i would use the cluster  for ioBroker (in that order)

1.  multus tooling to support addition networks (VLANs) for my ioBroker pods so they can receive the unicast/multicast traefik of the the devices in the isolated home-automation nework
2. Spread the load across nodes
3. Think about if i need HA

So HA is my second goal in this regard - i'am not sure 2 and 3 involve the same things, i assume you already have case 1 up and running, just without HA (and for me, that's already a huge win).

Did you already handle 1. or do you run it all in on single network? 

@Apollon77 

Those things are done in different ways with and without kubernetes. Without you usually have Qurums and a leader election and so forth, but i will not go into further details here. 

First i want to get it up and running securely, then getting out the value i need (this is a stage you all have passed already) and then think about spreading/HA. I am not even yet convinced that i want to get as much involved as writing an operator for ioBroker - so would just be honest about all of this. The only thing i def. will invest in is running it on k8 via a helm chart including a networking concept

**Reply by UncleSamSwiss** [reply]
@EugenMayer For the networking part I cheated: all adapters that need direct access to the network (either for server ports (HM RPC, ...) or broadcast/multicast) are running on my Raspi "slaves" and not in Kubernetes; the only exception being ioBroker.admin for which I added a LoadBalancer service. I'm also using a single Redis server for now (no cluster), so theoretically there is no master/slave, but most adapters are running in my k3s.

**Reply by Apollon77** [reply]
@EugenMayer I know how k8s works and in fact I wanted to state that ipbroker usees completely different "core concept" for the "hosts" and this is very deep into the "core". So it is not that easy to "simply change that". And yes the HA-Adapter ide was also not meaned to work ion k8s or such solutions but "how to make a classical iobroker with mutilpe hosts acting a bit more High available"

**Reply by EugenMayer** [reply]
Just as a heads up, i used multus with macvlan to attach my iobroker pod to an additional interface.

In this case, i 

- run a k3s vm (proxmox/qemu), with 2 interfaces, first is LAN, second is iot (VLAN). 
- Deploy a multus macvlan network like described here https://github.com/EugenMayer/helm-charts/blob/main/charts/iobroker/README.networking.md#macvlan
- annotate the pod as described above and i end up having the iobroker in both networks

When selecting adapters, i can properly select the interface to use, i use the iot interface and thats about it.

I did yet fully test multicast/unicast for device discovery, but i expect this to work fairly straight forward.

@Apollon77 
i'am (now) aware that main/minion mode is not meant to be for HA, but rather for horizontal scaling/isolation with manual load distribution (selecting instance when installing an adapter). I'am still interested in that, in terms that i would like to offload some of the adapters to isolated devices (for either of the reasons, isolation or load distribution). AFAICS i do not need to support this in the chart primarily, since switching a iobroker into minion mode is a in-software configuration. So to deploy this, all i need is running a proper k8s cluster and then support node-selection and install multiple iobroker charts into different namespaces, 1 for  main and then for each minion. So this is supported out of the box since it is all namespaced anyway.
